lib/Support/CFG.cpp

```diff
@@ -281,36 +281,64 @@ struct EndpointComparator {
   }
 };
 
-/// Define a map from BBEndpoints to the CFGEdges that connect the BBs
-using BBEndpointsMap =
-    std::map<BBEndpoints, llvm::DenseSet<CFGEdge *>, EndpointComparator>;
 } // namespace
 
+static BBEndpointsOptional getCFGEdgeEndpoints(const CFGEdge &edge) {
+  BBEndpointsOptional endpoints;
+
+  Operation *definingOp = edge.get().getDefiningOp();
+  if (!definingOp) {
+    endpoints.srcBB = std::nullopt;
+  } else {
+    if (auto bb = getLogicBB(definingOp)) {
+      endpoints.srcBB = *bb;
+    } else {
+      endpoints.srcBB = std::nullopt;
+    }
+  }
+
+  if (auto bb = getLogicBB(edge.getOwner())) {
+    endpoints.dstBB = *bb;
+  } else {
+    endpoints.dstBB = std::nullopt;
+  }
+
+  return endpoints;
+}
+
 BBtoArcsMap dynamatic::getBBPredecessorArcs(handshake::FuncOp funcOp) {
-  BBEndpointsMap endpointEdges;
```

Comment:
Since the key of `BBEndpointsMap` is now `BBEndpointsOptional`, which includes `std::optional<unsigned>`, this is not supported as a key in `std::map`.
As a result, I removed this type and updated the algorithm to avoid using it, which also simplifies the implementation.